table/block_based/block_based_table_reader.cc

```diff
@@ -1332,34 +1332,54 @@ Status BlockBasedTable::PrefetchIndexAndFilterBlocks(
     // Should we use FindOptionalMetaBlock here?
     s = FindMetaBlock(meta_iter, kUserDefinedIndexPrefix + udi_name,
                       &udi_block_handle);
-    if (!s.ok()) {
-      return s;
-    }
     // Read the block, and allocate on heap or pin in cache. The UDI block is
     // not compressed. RetrieveBlock will verify the checksum.
-    s = RetrieveBlock(prefetch_buffer, ro, udi_block_handle,
-                      rep_->decompressor.get(), &rep_->udi_block,
-                      /*get_context=*/nullptr, lookup_context,
-                      /*for_compaction=*/false, use_cache, /*async_read=*/false,
-                      /*use_block_cache_for_lookup=*/false);
+    if (s.ok()) {
+      s = RetrieveBlock(prefetch_buffer, ro, udi_block_handle,
+                        rep_->decompressor.get(), &rep_->udi_block,
+                        /*get_context=*/nullptr, lookup_context,
+                        /*for_compaction=*/false, use_cache,
+                        /*async_read=*/false,
+                        /*use_block_cache_for_lookup=*/false);
+    }
+    if (s.ok()) {
+      assert(!rep_->udi_block.IsEmpty());
+
+      std::unique_ptr<UserDefinedIndexReader> udi_reader =
+          table_options.user_defined_index_factory->NewReader(
+              rep_->udi_block.GetValue()->data);
+      if (udi_reader) {
+        index_reader = std::make_unique<UserDefinedIndexReaderWrapper>(
+            udi_name, std::move(index_reader), std::move(udi_reader));
+      } else {
+        s = Status::Corruption("Failed to create UDI reader");
+      }
+    }
     if (!s.ok()) {
-      return s;
+      RecordTick(rep_->ioptions.statistics.get(),
+                 SST_USER_DEFINED_INDEX_LOAD_FAIL_COUNT);
+      if (table_options.fail_if_no_udi_on_open) {
+        s = Status::Corruption("Failed to load UDI: " + udi_name, s.ToString());
```

Comment:
Minor: this overwrites the failure where the UDI block is available but creating a reader on it failed.

Reply:
The original status is captured in the string message. This way, we can provide some additional info such as the index name and file name.